src/cargo/sources/git/utils.rs

```diff
@@ -355,12 +365,33 @@ impl<'a> GitCheckout<'a> {
                     "Skipping",
                     format!(
                         "git submodule `{}` due to update strategy in .gitmodules",
-                        url
+                        child_url_str
                     ),
                 )?;
                 return Ok(());
             }
 
+            // See the `git submodule add` command on the documentation:
+            // https://git-scm.com/docs/git-submodule
+            let url = if child_url_str.starts_with("./") || child_url_str.starts_with("../") {
+                let mut new_path = Cow::from(parent_remote_url.path());
+                if !new_path.ends_with('/') {
+                    new_path.to_mut().push('/');
+                }
+
+                let mut new_parent_remote_url = parent_remote_url.clone();
+                new_parent_remote_url.set_path(&new_path);
```

Comment:
nit: I'd find this easier to follow if `new_path` was defined next to the use

Something like
```rust
                let mut new_parent_remote_url = parent_remote_url.clone();

                let mut new_path = Cow::from(parent_remote_url.path());
                if !new_path.ends_with('/') {
                    new_path.to_mut().push('/');
                }
                new_parent_remote_url.set_path(&new_path);
```